ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala

```diff
@@ -79,22 +80,29 @@ private[apiserver] final class ApiPackageManagementService private (
       .andThen(logger.logErrorsOnCall[ListKnownPackagesResponse])
   }
 
+  private[this] val darReader = DarReader { case (_, x) => Try(Archive.parseFrom(x)) }
+
+  def decodeAndValidate(stream: ZipInputStream): Try[Dar[Archive]] =
+    for {
+      dar <- darReader.readArchive("package-upload", stream)
+      packages <- Try(dar.all.iterator.map(Decode.decodeArchive).toMap)
+      _ <- Validation.checkPackages(packages).toTry
+    } yield dar
+
   override def uploadDarFile(request: UploadDarFileRequest): Future[UploadDarFileResponse] = {
     val submissionId =
       if (request.submissionId.isEmpty)
         SubmissionId.assertFromString(UUID.randomUUID().toString)
       else
         SubmissionId.assertFromString(request.submissionId)
 
+    val stream = new ZipInputStream(new ByteArrayInputStream(request.darFile.toByteArray))
```

Comment:
Do any of those streams need to be explicitly closed?

Reply:
Good question. It seems that `DarReader#readArchive` close the stream given in argument.

```scala

  /** Reads an archive from a ZipInputStream. The stream will be closed by this function! */
  def readArchive(
      name: String,
      darStream: ZipInputStream,
      entrySizeThreshold: Int = EntrySizeThreshold,
  ): Try[Dar[A]] = ...
```